src/Storages/MergeTree/DataPartsExchange.cpp

```diff
@@ -190,25 +199,46 @@ void Service::processQuery(const HTMLForm & params, ReadBuffer & /*body*/, Write
             writeBinary(projections.size(), out);
         }
 
+        const auto & part_storage = part->getDataPartStorage();
+        const String disk_type = part_storage.getDiskType();
+        const String disk_name = part_storage.getDiskName();
+
+        const Strings desired_zero_copy_disk_types = parseCommaDelimited(params.get(DISK_ZERO_COPY, ""));
         if (data_settings->allow_remote_fs_zero_copy_replication &&
-            /// In memory data part does not have metadata yet.
-            !isInMemoryPart(part) &&
-            client_protocol_version >= REPLICATION_PROTOCOL_VERSION_WITH_PARTS_ZERO_COPY)
+            !isInMemoryPart(part) && /// In memory data part does not have metadata yet.
+            client_protocol_version >= REPLICATION_PROTOCOL_VERSION_WITH_PARTS_ZERO_COPY &&
+            contains(desired_zero_copy_disk_types, disk_type))
         {
-            auto disk_type = part->getDataPartStorage().getDiskType();
-            if (part->getDataPartStorage().supportZeroCopyReplication() && std::find(capabilities.begin(), capabilities.end(), disk_type) != capabilities.end())
-            {
-                /// Send metadata if the receiver's capabilities covers the source disk type.
-                response.addCookie({"remote_fs_metadata", disk_type});
-                sendPartFromDisk(part, out, client_protocol_version, true, send_projections);
-                return;
-            }
+            response.addCookie({DISK_ZERO_COPY, disk_type});
+            sendPartFromDisk(part, out, client_protocol_version, RemoteDiskFeature::Zerocopy, send_projections);
+            return;
+        }
+
+        // TODO myrrc Unlike 0copy, the information about disk type (s3/hdfs/azure) is not sufficient to decide
+        // whether to send VFS metadata. 0copy operates on a table level, so metadata is just copied and
+        // later accessed. For VFS we have different logs for every disk, so if we download data from
+        // disk s1 to disk s2, we can't send metadata as otherwise logs would be corrupted.
+        // So currently VFS is limited. This has an implication that if different replicas have
+        // disks with same names but different endpoints (+credentials?), everything will break.
+        // A possible solution is to introduce a _vfs_disk_id_ parameter (by default a hash of
```

Comment:
> (+credentials?/ ... + key + secret)

Probably, wouldn't work for assumeRole or IAM provided credentials (basically when credentials consumed from environment and not created by users), which is "most" advisable and one of popular method of providing them. 

> _vfs_disk_id_

Can it be stored (and automatically generated) by clickhouse vfs on top level of object storage disk location? (maybe, it also can include some additional metadata in it, like migration status). It can be generated as random UUID. (probably only problem here is lack of PutObjectIfNotExist)